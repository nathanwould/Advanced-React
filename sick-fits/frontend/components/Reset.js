import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Form from './styles/Form';
import useForm from '../lib/useForm';
// import { CURRENT_USER_QUERY } from './User';
import Error from './ErrorMessage';

const RESET_MUTATION = gql`
  mutation RESET_MUTATION($email: String!, $token: String!) {
    redeemUserPasswordResetToken(email: $email, token: $token) {
      code
      message
    }
  }
`;

export default function Reset({ token }) {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
    token,
  });
  const [signup, { data, loading, error}] = useMutation(RESET_MUTATION, {
    variables: inputs,
    // refetch currently logged in user
    // refetchQueries: CURRENT_USER_QUERY,
  });

  const successfulError = data?.redeemUserPasswordResetToken?.code
    ? data?.redeemUserPasswordResetToken
    : undefined;
  async function handleSubmit(e) {
    e.preventDefault();
    console.log(inputs);
    const res = await signup();
    console.log(res);
    resetForm();
  }
  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Reset Your Password</h2>
      <Error error={error || successfulError} />
      <fieldset>
        {data?.redeemUserPasswordResetToken === null && (
          <p>Success! You can now sign in</p>
        )}
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Your Email Address"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Request Reset</button>
      </fieldset>
    </Form>
  );
}
