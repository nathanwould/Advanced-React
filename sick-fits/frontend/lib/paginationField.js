import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    // tells Apollo we will take care of everything
    keyArgs: false,
    read(existing = [], { args, cache }) {
      // console.log({ existing, args, cache });
      const { skip, first } = args;

      // read the number of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // Check if we have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);

      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        return false;
      }
      if (items.length) {
        // console.log(
        //   `There are ${items.length} items in the cache! Gonna send them to Apollo!`
        // );
        return items;
      }
      // fallback to network
      return false;
    },
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; i++) {
        merged[i] = incoming[i - skip];
      }
      return merged;
      // console.log(args)
    },
  };
}
