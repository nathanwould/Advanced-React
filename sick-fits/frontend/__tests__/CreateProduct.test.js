import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import Router from 'next/router';
import CreateProduct, {
  CREATE_PRODUCT_MUTATION,
} from '../components/CreateProduct';
import { fakeItem, makePaginationMocksFor } from '../lib/testUtils';
import { ALL_PRODUCTS_QUERY } from '../components/Products';

const item = fakeItem();

describe('<CreateProduct />', () => {
  it('renders and matches snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it('handles the updating', async () => {
    // render the form out
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    // type into the boxes
    await userEvent.type(screen.getByPlaceholderText(/name/i), item.name);
    // await userEvent.type(
    //   screen.getByPlaceholderText(/price/i),
    //   item.price.toString()
    // );
    await userEvent.type(
      screen.getByPlaceholderText(/description/i),
      item.description
    );
    debug();
    // check that those boxes are populated
    expect(screen.getByDisplayValue(item.name)).toBeInTheDocument();
    // expect(screen.getByDisplayValue(item.price)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.description)).toBeInTheDocument();
  });
});
