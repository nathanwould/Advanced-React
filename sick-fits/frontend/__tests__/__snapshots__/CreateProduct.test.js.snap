// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct /> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 hVEEdl"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="image"
      >
        Image
        <input
          id="image"
          name="image"
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="name"
          type="text"
          value=""
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="price"
          type="number"
          value="0"
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="description"
        />
      </label>
      <button
        type="submit"
      >
        + Add Product
      </button>
      <button
        type="button"
      >
        Clear Form
      </button>
      <button
        type="button"
      >
        Reset Form
      </button>
    </fieldset>
  </form>
</div>
`;
